doc/byte_stream.md

```diff
@@ -0,0 +1,137 @@
+# Byte Streams
+
+***WARNING*** this is currently a proposal that is not implemented yet.
+
+This document describes a mechanism by which OpenMRN components can exchange a
+unidirectional untyped data stream. The API is laid out for compatibility with
+the StateFlow concept and provides asynchronous implementations to limit memory
+usage and flow control.
+
+## Use-cases
+
+- OpenLCB Stream sending. The client component sends data to the stream
+  transmitter using Byte Streams.
+  
+- OpenLCB Stream receiving. The client component is receiving data from the
+  Stream service using Byte Streams.
+  
+- A fast implementation of a TCP Hub should use Byte Streams to represent the
+  TCP messages for proxying between ports.
+  
+## Non-goals
+
+- It is not a goal to use the Byte Stream API to write from a State Flow to an
+  fd. (Neither sockets, pipes, nor physical files.) Writing to an fd should be
+  done by the native StateFlow features like `write_repeated`.
+
+## Requirements
+
+- Memory allocation
+
+  - Allocation has to avoid memory fragmentation. While the expectation is that
+    larger blocks of memory are allocated in one go, these blocks should be
+    reused between different users of Byte Streams and at different times,
+    instead of returning to malloc/free.
+  
+  - Block allocation size should be 1 kbyte. This is small enough that even
```

Comment:
What about making the allocation size exactly 1 MTU, where MTU is configurable for the use case?

Reply:
Making the allocation size a compile time constant (constexpr or `#define`) seems plausible. Making it a run-time constant is a lot more difficult, and probably also a link-time constant is hard. There are places in openmrn that are doing `sizeof(Buffer<T>)` and those will fail when T's size is not a compile time constant.

Reply:
I understand. A constexpr seems fine to me.